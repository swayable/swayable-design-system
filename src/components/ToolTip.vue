<template>
  <component
    :is='type'
    ref='trigger'
    class='relative'
    :class='`tooltip-${position}-wrapper`'
    @mouseover='onMouseover'
    @mouseleave='onMouseleave'
  >
    <slot />
    <div
      v-show='show'
      :style='crossAxisPosition'
      :class='`tooltip-${position}`'
      class='absolute z-50'
    >
      <div
        class='tooltip text-white rounded-lg py-1 px-2 whitespace-no-wrap'
        :class='classNames'
      >
        <slot name='tip' />
      </div>
    </div>
  </component>
</template>

<script>
const STATE = {
  closed: 0,
  opening: 1,
  open: 2,
}

export default {
  name: 'ToolTip',
  status: 'ready',
  props: {
    /**
     * The html element used.
     */
    type: {
      type: String,
      default: 'div',
    },
    /**
     * Tooltip position in relation to the element
     */
    position: {
      type: String,
      default: 'top',
      validator: (val) => [
        'top',
        'bottom',
        'left',
        'right',
      ].includes(val),
    },
    /**
     * Causes tooltip to open aligned with the cursor. Otherwise is centered.
     */
    cursorAlign: {
      type: Boolean,
      default: false,
    },
    /**
     * Milliseconds between mouseover and opening
     */
    delay: {
      type: Number,
      default: 200,
    },
    /**
     * Class names are applied to the tooltip
     */
    classNames: {
      type: String,
      default: '',
    },
  },
  data() {
    return {
      state: STATE.closed,
      crossAxisPosition: {},
    }
  },
  computed: {
    show() {
      return this.state === STATE.open
    },
  },
  methods: {
    onMouseover(e) {
      if (this.state === STATE.closed) this.state = STATE.opening
      setTimeout(() => {
        if (this.state === STATE.opening) {
          if (this.cursorAlign) this.alignWithCursor(e)
          this.state = STATE.open
        }
      }, this.delay)
    },
    onMouseleave() {
      this.state = STATE.closed
    },
    alignWithCursor({ clientX, clientY }) {
      const { left, top } = this.$refs.trigger.getBoundingClientRect()
      const positionTopOrBottom = ['top', 'bottom'].includes(this.position)

      this.crossAxisPosition = positionTopOrBottom
        ? { left: `${clientX - left}px`}
        : { top: `${clientY - top}px`}
    },
  },
}
</script>

<style lang="scss">
// TODO: dark mode requires some refactoring, but is so-so ok for now
$tooltipColor: adjust-color($color-dark-3, $alpha: -0.1);
.tooltip {
  background: $tooltipColor;
  position: relative;
  &:before {
    content: ' ';
    width: 0;
    height: 0;
    border: theme('spacing.1') solid transparent;
    position: absolute;
  }
}

.tooltip-top {
  bottom: calc(100% - theme('spacing.1'));
  left: 50%;
  transform-origin: bottom left;
  transform: translateX(-50%) translateY(0%);
  .tooltip {
    bottom: theme('spacing.2');
    &:before {
      top: 100%;
      left: calc(50% - theme('spacing.1'));
      border-top-color: $tooltipColor;
    }
  }
}
.tooltip-bottom {
  top: calc(100% - theme('spacing.1'));
  left: 50%;
  transform-origin: top left;
  transform: translateX(-50%) translateY(0%);
  .tooltip {
    top: theme('spacing.2');
     &:before {
      bottom: 100%;
      left: calc(50% - theme('spacing.1'));
      border-bottom-color: $tooltipColor;
    }
  }
}
.tooltip-left {
  top: 50%;
  right: calc(100% - theme('spacing.1'));
  transform-origin: top right;
  transform: translateX(0%) translateY(-50%);
  .tooltip {
    right: theme('spacing.2');
    &:before {
      left: 100%;
      top: calc(50% - theme('spacing.1'));
      border-left-color: $tooltipColor;
    }
  }
}
.tooltip-right {
  top: 50%;
  left: calc(100% - theme('spacing.1'));
  transform-origin: top left;
  transform: translateX(0%) translateY(-50%);
  .tooltip {
    left: theme('spacing.2');
    &:before {
      right: 100%;
      top: calc(50% - theme('spacing.1'));
      border-right-color: $tooltipColor;
    }
  }
}
</style>

<docs>
  ```jsx
  <div class='flex justify-around mb-5 theme-dark-mode'>
    <ToolTip position='right'>
      <Label>Right</Label>
      <template #tip>
        <p>Tooltip Right</p>
      </template>
    </ToolTip>
    <ToolTip position='bottom'>
      <Label>Bottom</Label>
      <template #tip>
        <p>Tooltip Bottom</p>
      </template>
    </ToolTip>
    <ToolTip position='left'>
      <Label>Left</Label>
      <template #tip>
        <p>Tooltip Left</p>
      </template>
    </ToolTip>
    <ToolTip position='top'>
      <Label>Top</Label>
      <template #tip>
        <p>Tooltip Top</p>
      </template>
    </ToolTip>
    <ToolTip :delay='500' position='bottom'>
      <Label>Delayed</Label>
      <template #tip>
        <p>by half a second</p>
      </template>
    </ToolTip>
    <ToolTip :delay='0' position='bottom'>
      <Label>Not Delayed</Label>
      <template #tip>
        <p>at all</p>
      </template>
    </ToolTip>
  </div>
  ```
</docs>
