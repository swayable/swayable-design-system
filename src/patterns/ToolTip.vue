<template>
  <component
    :is='type'
    class='relative'
    :class='`tooltip-${position}-wrapper`'
  >
    <span
      @mouseover='show = true'
      @mouseleave='show = false'
    >
      <slot />
    </span>
    <transition name='expand'>
      <div
        v-show='show'
        class='tooltip absolute bg-dark text-white z-40 rounded text-center py-1 px-2 whitespace-no-wrap'
        :class='`tooltip-${position}`'
      >
        <slot name='tip' />
      </div>
    </transition>
  </component>
</template>

<script>
const bottom = { top: '100%', left: 'auto', right: '50%', transform: 'translateX(50%)'  }
const top = { bottom: '100%', left: 'auto', right: '50%', transform: 'translateX(50%)' }
const right = { left: '100%', top: 'auto', bottom: '50%', transform: 'translateY(50%)' }
const left = { right: '100%', top: 'auto', bottom: '50%', transform: 'translateY(50%)'  }
const anchorMap = {
  top,
  bottom,
  left,
  right,
}

export default {
  name: 'ToolTip',
  status: 'prototype',
  release: '0.1.0',
  props: {
    /**
       * The html element used.
       */
    type: {
      type: String,
      default: 'div',
    },
    /**
       * The html element used.
       */
    position: {
      type: String,
      default: 'bottom',
      validator: (val) => Object.keys(anchorMap).includes(val),
    },
  },
  data() {
    return { show: false }
  },
}
</script>

<style lang="scss">
.tooltip:after {
  content: ' ';
  position: absolute;
  background: inherit;
  width: 10px;
  height: 10px;
  transform: rotate(45deg);
  z-index: -1;
}
.tooltip-top-wrapper, .tooltip-bottom-wrapper {
  .expand-enter-active { animation: expandY .2s }
  .expand-leave-active { animation: expandY .2s reverse }
}
@keyframes expandY {
  0% { transform: scale(0) translateX(-50%) translateY(0%); }
  50% { transform: scale(1.1) translateX(-50%) translateY(0%); }
  100% { transform: scale(1) translateX(-50%) translateY(0%); }
}

.tooltip-left-wrapper, .tooltip-right-wrapper {
  .expand-enter-active { animation: expandX .2s }
  .expand-leave-active { animation: expandX .2s reverse }
}
@keyframes expandX {
  0% { transform: scale(0) translateX(0%) translateY(-50%); }
  50% { transform: scale(1.1) translateX(0% translateY(-50%)); }
  100% { transform: scale(1) translateX(0%) translateY(-50%); }
}

.tooltip-top {
  bottom: calc(100% + 4px);
  left: 50%;
  transform-origin: bottom left;
  transform: translateX(-50%) translateY(0%);
  &:after { top: calc(100% - 8px) }
}
.tooltip-bottom {
  top: calc(100% + 4px);
  left: 50%;
  transform-origin: top left;
  transform: translateX(-50%) translateY(0%);
  &:after { bottom: calc(100% - 8px) }
}
.tooltip-left {
  top: 50%;
  right: calc(100% + 4px);
  transform-origin: top right;
  transform: translateX(0%) translateY(-50%);
  &:after {
    top: calc(50% - 4px);
    left: calc(100% - 8px);
  }
}
.tooltip-right {
  top: 50%;
  left: calc(100% + 4px);
  transform-origin: top left;
  transform: translateX(0%) translateY(-50%);
  &:after {
    top: calc(50% - 4px);
    right: calc(100% - 8px);
  }
}
</style>

<docs>
  ```jsx
  <div class='flex justify-around mb-5'>
    <ToolTip position='right'>
      <p class='py-1 px-2 rounded border'>Right</p>
      <template #tip>
        <p>Tooltip Right</p>
      </template>
    </ToolTip>
    <ToolTip position='bottom'>
      <p class='py-1 px-2 rounded border'>Bottom</p>
      <template #tip>
        <p>Tooltip Bottom</p>
      </template>
    </ToolTip>
    <ToolTip position='left'>
      <p class='py-1 px-2 rounded border'>Left</p>
      <template #tip>
        <p>Tooltip Left</p>
      </template>
    </ToolTip>
    <ToolTip position='top'>
      <p class='py-1 px-2 rounded border'>Top</p>
      <template #tip>
        <p>Tooltip Top</p>
      </template>
    </ToolTip>
  </div>
  ```
</docs>
