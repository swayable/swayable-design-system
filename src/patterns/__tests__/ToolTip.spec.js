import { shallowMount } from '@vue/test-utils'
import ToolTip from '../ToolTip'

describe('ToolTip.vue', () => {
  describe('position top', () => {
    it('matches the snapshot', () => {
      const wrapper = shallowMount(ToolTip, { propsData: { position: 'top' } })
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('position right', () => {
    it('matches the snapshot', () => {
      const wrapper = shallowMount(ToolTip, { propsData: { position: 'right' } })
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('position bottom', () => {
    it('matches the snapshot', () => {
      const wrapper = shallowMount(ToolTip, { propsData: { position: 'bottom' } })
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('position left', () => {
    it('matches the snapshot', () => {
      const wrapper = shallowMount(ToolTip, { propsData: { position: 'left' } })
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})
