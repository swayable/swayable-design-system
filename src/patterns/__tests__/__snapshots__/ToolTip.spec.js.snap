// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolTip.vue position bottom matches the snapshot 1`] = `
<div class="relative tooltip-bottom-wrapper">
  <div class="tooltip text-white absolute z-50 rounded-lg py-1 px-2 whitespace-no-wrap tooltip-bottom undefined" style="display: none;" name="expand"></div>
</div>
`;

exports[`ToolTip.vue position left matches the snapshot 1`] = `
<div class="relative tooltip-left-wrapper">
  <div class="tooltip text-white absolute z-50 rounded-lg py-1 px-2 whitespace-no-wrap tooltip-left undefined" style="display: none;" name="expand"></div>
</div>
`;

exports[`ToolTip.vue position right matches the snapshot 1`] = `
<div class="relative tooltip-right-wrapper">
  <div class="tooltip text-white absolute z-50 rounded-lg py-1 px-2 whitespace-no-wrap tooltip-right undefined" style="display: none;" name="expand"></div>
</div>
`;

exports[`ToolTip.vue position top matches the snapshot 1`] = `
<div class="relative tooltip-top-wrapper">
  <div class="tooltip text-white absolute z-50 rounded-lg py-1 px-2 whitespace-no-wrap tooltip-top undefined" style="display: none;" name="expand"></div>
</div>
`;
